.github/workflows/release.yml

```diff
@@ -20,32 +20,19 @@ jobs:
       - test
       - packages
     runs-on: ubuntu-latest
-    env:
-      PYPI_SECRET_PATH: secret/apm-team/ci/apm-agent-python-pypi-prod
+    environment: release
+    permissions:
+      id-token: write  # IMPORTANT: this permission is mandatory for trusted publishing
     steps:
       - uses: actions/checkout@v3
-      - uses: hashicorp/vault-action@v2.4.2
-        with:
-          url: ${{ secrets.VAULT_ADDR }}
-          method: approle
-          roleId: ${{ secrets.VAULT_ROLE_ID }}
-          secretId: ${{ secrets.VAULT_SECRET_ID }}
-          secrets: |
-            ${{ env.PYPI_SECRET_PATH }} user | TWINE_USER ;
-            ${{ env.PYPI_SECRET_PATH }} password | TWINE_PASSWORD
       - uses: actions/download-artifact@v3
         with:
           name: packages
           path: dist
       - name: Upload
-        run: |
-          python -m pip install --user twine
-          python setup.py sdist
-          echo "Uploading to ${REPO_URL} with user ${TWINE_USER}"
-          python -m twine upload --username "${TWINE_USER}" --password "${TWINE_PASSWORD}" --skip-existing --repository-url ${REPO_URL} dist/*.tar.gz
-          python -m twine upload --username "${TWINE_USER}" --password "${TWINE_PASSWORD}" --skip-existing --repository-url ${REPO_URL} dist/*.whl
-        env:
-          REPO_URL: "https://upload.pypi.org/legacy/"
+        uses: pypa/gh-action-pypi-publish@f5622bde02b04381239da3573277701ceca8f6a0
```

Comment:
I think we still need a step to run `python setup.py sdist`. I think it makes most sense to add it here: https://github.com/elastic/apm-agent-python/blob/main/.github/workflows/packages.yml#L26

Reply:
Does it make sense since We are already relying on this step to generate artifacts and then downloading those artifacts before upload?

Reply:
exactly. And since `sdist` generates a source distribution package, it makes sense to do it in the `packages.yml` workflow. I'm not sure why we did it here to begin with, probably it was just copy&pasted from the old Jenkins setup.

Reply:
Ah, I got it. You mean to download the artifacts already upload in the other workflow rather than running the step within the same workflow?
I guess it because there are extra steps to download artifacts from another workflow.

Reply:
Don't we already do that? The whl file we generate in the packages workflow is uploaded in the release workflow AFAIK

Reply:
Yeah, I don't care where those steps are, but we need to run `python setup.py sdist` somewhere. If you look at this [doc](https://docs.pypi.org/trusted-publishers/using-a-publisher/), it's the `# retrieve your distributions here` piece of the puzzle. We need to generate those packages somewhere so that the publish action can publish them.

Reply:
Currently, we have a job `packages` within the workflow `release` that executes the `python setup.py sdist`.
The `publish-pypi` download the result of the `packages` job which is a list of distribution package.
AFAIU, the `python setup.py sdist` wasn't needed because it was already performed in another job within the `release` workflow.

In our case (within the `release` workflow), we can remove the `python setup.py sdist` step from the `publish-pypi` job and rely on the `packages` job to do it or rely on the `publish-pypi` job to do it and remove the `packages` job.
The first option avoid duplicate CI code since those steps are already present in the `packages` workflow definition.

> Ah, I got it. You mean to download the artifacts already upload in the other workflow rather than running the step within the same workflow?
> I guess it because there are extra steps to download artifacts from another workflow.
In some case, there is also another option which is to download the result of the `build` job from the `packages` workflow.

Reply:
Just to make sure we are on the same page: currently, there are two release artifacts that need to be built and pushed to PyPI:

 * a source distribution, `.tar.gz`, generated by `python setup.py sdist`
 * a "universal wheel" distribution, `.whl`, generated by `python setup.py bdist_wheel` (this command needs the `wheel` package to be installed during the build)

We used to generate a lot more wheel packages for all kinds of different platforms, but since we removed any platform-dependent code, only one universal wheel is needed. Because the package build process was much more complicated in the past, it was split into a separate job, and run on every commit.

I guess since the package building is much simpler these days, we could remove the `packages.yml` job completely and only build the packages in the release workflow.